clients/lsp-javascript.el

```diff
@@ -790,20 +790,42 @@ name (e.g. `data' variable passed as `data' parameter)."
   (when-let ((workspace (lsp-find-workspace 'ts-ls (buffer-file-name))))
     (eq 'initialized (lsp--workspace-status workspace))))
 
-(defun lsp-clients-typescript-project-ts-server-path ()
-  "Return the project local TS server path."
-  (f-join (lsp-workspace-root) "node_modules" "typescript" "lib" "tsserver.js"))
+(defun lsp-clients-typescript-require-resolve (&optional dir)
+  "Get the location of the typescript.
+Use Node.js require.
+The node_modules directory structure is suspect
+and should be trusted as little as possible.
+If you call require in Node.js,
+it should take into account the various hooks.
+For example, yarn PnP.
+
+Optional argument DIR specifies the working directory
+to run the command in."
+  (let* ((default-directory (or dir default-directory))
+         (output
+          (string-trim-right
+           (shell-command-to-string
+            "node -e \"console.log(require.resolve('typescript'))\""))))
+    (if (string-empty-p output)
+        nil
+      (f-parent output))))
```

Comment:
This can also be simplified with `if-let*` or `when-let*`.

Reply:
I fixed.